### PR TITLE
Add auth options :tada:

### LINKED ISSUE #37: Authorisation
We have decided on authentication in #23. We'll now need to think about authorisation.

Authentication can be done using Auth0. Both /graphql and the graphcool IDE endpoints should be approachable without authentication, and only queries and mutations needing authentication should require it.

## User groups ##

* unauthenticated users -this might not be available for a while or at all
* 3rd parties acting on behalf of users
* 3rd parties authenticating with API keys
* logged in users, most likely via the site

## Authorisation ##

There are plenty of options here as the specification does not dictates a solution.

* Business logic layer: the recommendation of the GraphQL best practices doco (http://graphql.org/learn/authorization/)
* Chaining resolvers, see https://github.com/thebigredgeek/apollo-resolvers and the corresponding talk at Apollo Day https://www.youtube.com/watch?v=xaorvBjCE7A
* GraphQL shield is a project separating permission logic from the application logic allowing you to pass your resolvers and a permission object into it's API https://github.com/maticzav/graphql-shield
* Directives, see https://blog.graph.cool/graphql-directive-permissions-authorization-made-easy-54c076b5368e based on a talk by Ryan Chenkie at GraphQL summit: https://www.youtube.com/watch?v=4_Bcw7BULC8
#25 

A related thread full of interesting viewpoints https://github.com/apollographql/graphql-tools/issues/313

### PR DESCRIPTION
**Pull request by Bouncey**
Closes #37 
Closes #83

### CODE REVIEW

**Comment by raisedadead** [comment]
Quick dumb query, how do I set a JWT for testing?

**Comment by ojongerius** [comment]
Nice one 🙇 I'll dedicated some time to review this evening.

**Comment by Bouncey** [comment]
My bad, it was late at night.

## You can

### Use one of my expired tokens:
```sh
eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik0wTTJOVU0yTWpaR1FqUkVRelV4TVRFME9VTTVOVGcwUmtFeE1rTTJOek5GUTBRelJUTkVOZyJ9.eyJuaWNrbmFtZSI6InN0dWFydCIsIm5hbWUiOiJzdHVhcnRAZnJlZWNvZGVjYW1wLm9yZyIsInBpY3R1cmUiOiJodHRwczovL3MuZ3JhdmF0YXIuY29tL2F2YXRhci85ZjhmY2YzY2MzN2QzYWY4NDg0NDg3OTk4OTBmZGRmZT9zPTQ4MCZyPXBnJmQ9aHR0cHMlM0ElMkYlMkZjZG4uYXV0aDAuY29tJTJGYXZhdGFycyUyRnN0LnBuZyIsInVwZGF0ZWRfYXQiOiIyMDE4LTA0LTIyVDEwOjA5OjEzLjkzN1oiLCJlbWFpbCI6InN0dWFydEBmcmVlY29kZWNhbXAub3JnIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImlzcyI6Imh0dHBzOi8vZnJlZWNvZGVjYW1wLmF1dGgwLmNvbS8iLCJzdWIiOiJlbWFpbHw1YWNiMmVhZGE2ODAzYTkxOTEyODgwMTciLCJhdWQiOiJ2RjcwQ0paeVBLYlpSNHkwYXZlY1hYTGtmeU1ObnlLbiIsImlhdCI6MTUyNDM5MTc1MywiZXhwIjoxNTI0NDI3NzUzLCJhdF9oYXNoIjoiZDZsSW9Ra19uWkwxakxGWlFNN01EdyIsIm5vbmNlIjoiMkJWbVlWOWkzckllSFVJZzhzUU5WTDZtXy1MZTREejUifQ.nw5kbnpqCZRrJN3qr9coUAMNCRRg3Vrs7jhSqyX_BqGNcxZRA0ZYovM0ZNHQBioy_XH0bEeIoy59Bgjzcd-hNFJyoa_8pXRbhJ4tBMYkdK7K8_d8rGdDdSDg2EkTTeeidH8uAx9E14nYCWUJdqRlD1YXp3XkqQtBPLkD5Fu296ZnGrZ8hCjqz78bvYEg-dXYd3NsuuplnbFTjhjhHJ57KxBvk6_2Xvg-eeNx73xOJBsscyStv6eVvzQKMLvEin7SbV4FA4C0cKlKZl7YVj2qVxehojxkDupnSxFCDb59gkf4Z99eOx8qQOE4lI5c3gKuoZ3sIYDd1aRgHXAsvl8Dag
```

Change [this line](https://github.com/Bouncey/open-api/blob/38c192d14e5a9e4e3c66ae54c2a4c71ab3f5ef3d/graphql/resolvers/directives.js#L22) to 
```js
 return jwt.verify(token.replace('Bearer ', ''), cert, { ignoreExpiration: true });
```
### Or generate your own by:

1. run `events` locally
1. log in
1. access `id_token` from your `localStorage` 

## So, you have a token

### Add your token to the request headers :point_down: 

headers


Play around with the directives on the `User` type and query. 

The Idea is:
- if a field has the `@isAuthenticatedOnField` directive, the value will be `null` if they are not authenticated, but other information in the request will be available.
- if a query has the `@isAuthenticatedOnQuery` directive, the request will error out if they are not authenticated.


I hope this helps in your QA @raisedadead @ojongerius  :grinning:

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js]: If we expect this file to be there, we'll need to make sure it's available for CI to deploy too.

Another thought: if we pull this out of the hander, this will persist across Lambda invocations and we won't have to reload the file. Like I did for the database connection logic in https://github.com/freeCodeCamp/open-api/blob/staging/handler.js#L13

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js]: It would be a courtesy to our users and future selves to tell them if their token has expired or something else went wrong. Glancing at [JsonWebTokenError](https://github.com/auth0/node-jsonwebtoken#jsonwebtokenerror) it could be fine to bubble those up to the user?

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js]: Should we just call this `validateWebtoken` or something else that is more descriptive of what we are actually doing here?

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js]: Is this still needed after https://github.com/apollographql/graphql-tools/pull/518 was merged?

**Review by ojongerius — CHANGES_REQUESTED** [review]
Nice work @Bouncey!

File: graphql/resolvers/directives.js
Comment (Bouncey) [reply graphql/resolvers/directives.js]: I didn't even realise this was a thing inside `graphql-tools`, I will update the PR to let `graphql-tools` worry about the implementation

File: graphql/resolvers/directives.js
Comment (Bouncey) [reply graphql/resolvers/directives.js]: Sounds reasonable

File: graphql/resolvers/directives.js
Comment (Bouncey) [reply graphql/resolvers/directives.js]: I didn't think about deployment when making changes, I'll scratch my head some more

File: graphql/resolvers/directives.js
Comment (Bouncey) [reply graphql/resolvers/directives.js]: Of course.

File: graphql/resolvers/directives.js
Comment (Bouncey) [inline graphql/resolvers/directives.js]: Still working on this...

File: .gitignore
Comment (ojongerius) [inline .gitignore]: Can be removed now that we are using env variables?

File: README.md
Comment (ojongerius) [inline README.md]: Can be removed now that we are using env variables?

File: README.md
Comment (ojongerius) [inline README.md]: Can be removed now that we are using env variables?

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js]: How about we just have `JWT_CERT`, and our deploy script figures out which variables are used, like we do for `MONGODB_URL` and `GRAPHQL_ENDPOINT_URL`? If we do that we'll need to edit `scripts/deploy.sh`

In any case you'll still need to add the env variables to `serverless.yml`.

File: graphql/resolvers/directives.js
Comment (ojongerius) [inline graphql/resolvers/directives.js:27]: 👍

File: public.pem.sample
Comment (ojongerius) [inline public.pem.sample]: Can go?

**Review by ojongerius — CHANGES_REQUESTED** [review]

**Review by raisedadead — CHANGES_REQUESTED** [review]
`package-lock.json` should be deleted. 

I would recommend this just to be sure:

1. `rm -rf node_modules package-lock.json`
2. `yarn`

do the git thing!

File: package.json
Comment (raisedadead) [inline package.json]: This whole line is not safe for cross systems. Here is an example of the behaviour on Windows (git bash for Windows):

```
$ yarn start
yarn run v1.6.0
$ DEBUG='fcc:*' nodemon ./node_modules/serverless/bin/serverless offline start --skipCacheInvalidation
'DEBUG' is not recognized as an internal or external command,
operable program or batch file.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: package.json
Comment (raisedadead) [reply package.json]: I would rather use a task runner like gulp to offload the heavy-lifting from here.

Comment (raisedadead) [reply package.json]: We can have this later and separate of this PR. Lets just get the package-lock removed and we are good to go.

**Review by ojongerius — APPROVED** [review]

File: package.json
Comment (raisedadead) [reply package.json]: HOLD on... I do not have `nodemon` globally on a separate *NIX system and I am getting this:

```
❯ yarn start
yarn run v1.6.0
(node:81582) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
$ DEBUG='fcc:*' nodemon ./node_modules/serverless/bin/serverless offline start --skipCacheInvalidation
/bin/sh: nodemon: command not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

on macOS

I think this is not expected right?

Comment (ojongerius) [reply package.json]: @raisedadead scratched his own itch and raised https://github.com/freeCodeCamp/open-api/pull/78

**Comment by ojongerius** [comment]
@raisedadead are you happy for this to go in? Mind you this will need conflicts resolved and I hope my tests will start failing 🤪

**Comment by ojongerius** [comment]
```sh
▶ yarn test
yarn run v1.6.0
$ jest --runInBand
 FAIL  test/user.test.js
  ✕ should return a user after one has been created (74ms)

  ● should return a user after one has been created

    TypeError: Cannot read property '0' of null

      40 |
      41 |   /* eslint-disable no-undef */
    > 42 |   expect(data.users[0].name).toBe(user.name);
      43 |   /* eslint-enable no-undef */
      44 | });
      45 |

      at Object.<anonymous>.it (test/user.test.js:42:10)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        1.394s, estimated 2s
Ran all test suites.
error Command failed with exit code 1.
```



When having a gander at what we do get back I see:
```js
    { errors:
       [ GraphQLError {
           message: 'Cannot read property \'headers\' of undefined',
           locations: [Array],
           path: [Array] } ],
      data: { users: null } }
```

Looks like the contents of `data` are all good, but the error object could to with some attention, and, as expected, we need to adjust these tests to validate requests with, and without authentication and authorisation behave as expected.

For the record, the same code using /api in a browser does return exactly what we want when not authenticated 👍 

```js
{
  "data": {
    "users": null
  },
  "errors": [
    {
      "message": "You must supply a JSON Web Token for authorization!",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "users"
      ]
    }
  ]
} 
```

File: graphql/resolvers/user.js
Comment (ojongerius) [inline graphql/resolvers/user.js:8]: `createUser`, no longer `dbUsers.createUser`?

File: handler.js
Comment (ojongerius) [inline handler.js]: I do not know why but if I run this locally I get:

```sh
[ 'Error: Serverless-offline: handler for \'graphql\' is not a function',
```

It works fine when I change it to:

```js
exports.graphqlHandler = async function (event, context, callback) {
```

File: handler.js
Comment (ojongerius) [inline handler.js]: Like above, works with `exports.apiHandler` but this give me a `handler for `api` is not a function`

**Review by ojongerius — CHANGES_REQUESTED** [review]

**Comment by Bouncey** [comment]
I did a `git pull --rebase upstream staging` which must have must have reverted things and I didn't notice. I'll untangle this mess when I get in to work

File: package.json
Comment (Bouncey) [reply package.json]: Nodemon was installed as a dep. I think the rebase after the switch to yarn messed things up. Will fix

File: handler.js
Comment (Bouncey) [reply handler.js]: I will revert

File: graphql/resolvers/user.js
Comment (Bouncey) [reply graphql/resolvers/user.js:8]: Sure, I am touching that in my user type PR.

**Comment by Bouncey** [comment]
I will take a look at the test suite. We may need to spilt the current tests in to smaller units, as currently it is a bit end-to-end -ish. The errors headers tells me that context hasn't been set in the handler correctly, for example.

**Comment by raisedadead** [comment]
Hi @ojongerius @Bouncey I just rebase and force pushed staging after #69 was merged. 

@ojongerius For all intents and purposes, its best to rebase an pr against staging, to avoid nasty merge commits. Or the escape hatch is to simple use squash commit button.

File: package.json
Comment (raisedadead) [reply package.json]: Well I must admit that OCD with DX will be a thing around here! 🤷‍♂️

Comment (ojongerius) [reply package.json]: Love it.

**Comment by Bouncey** [comment]
I want to see if I can get gulp to handle the tests due to the exporting of the `JWT_CERT` during testing

**Comment by Bouncey** [comment]
@raisedadead after rebasing and merging #83 into this PR I kept on getting this weird gulp error before I even touched the `gulpfile`
```sh
AssertionError [ERR_ASSERTION]: Task function must be specified
```
I have replaced what gulp was doing for us with `cross-env`. Can you please check everything is OK on your end using Windows?

**Comment by raisedadead** [comment]
Hi @Bouncey 

My primary reason for having a task runner was to have the ability to offload stuff from package.json. We could potentially use tasks for deployment and release (github, npm) automation instead of having raw shell files and workarounds in the run scripts.

But, I am OK with removing gulp as long as we are moving forward for now.

**Comment by Bouncey** [comment]
Sweet!

This is now ready for QA @raisedadead @ojongerius

File: auth/index.js
Comment (raisedadead) [inline auth/index.js]: `NODE_ENV === 'test'` => `NODE_ENV === 'development'`

or even better

`NODE_ENV !== 'production'`

Having it as 'test' which is a non-standard value will make the express server crash. This is because the third-party deps we use, such as playground rely on `NODE_ENV`

Mutating it to a non-standard value a bad idea IMHO.

**Review by raisedadead — CHANGES_REQUESTED** [review]

**Comment by Bouncey** [comment]
Oops! That shouldn't  be there now.

**Comment by raisedadead** [comment]
Other than that everything else LGTM. Go to merge.

Side note, I have to lookup but the headers variable like `authorization` should be capitalised?

`authorization` => `Authorization`

As a practice else where?

**Comment by Bouncey** [comment]
The headers are passed to us from serverless through context. We have no control over that. 

I was using `Authorization` in the playground headers, and they were always showing up as `authorization` in the handler context

**Comment by raisedadead** [comment]
Ok perfect.

**Review by raisedadead — APPROVED** [review]

**Comment by Bouncey** [comment]
Travis is failing on commitlint-travis. There is an open issue to relax these rules #54, all tests pass

**Comment by Bouncey** [comment]
Sweet!

**Pull request merged**
